File: Extensions/Tessa.Extensions.Default.Shared/Workflow/KrProcess/Formatters/ForkStageTypeFormatterBase.cs
#nullable enable

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Tessa.Cards;
using Tessa.Platform.Storage;

namespace Tessa.Extensions.Default.Shared.Workflow.KrProcess.Formatters
{
    /// <summary>
    /// Базовый абстрактный класс форматтера этапов типов "Ветвление".
    /// </summary>
    public abstract class ForkStageTypeFormatterBase :
        StageTypeFormatterBase
    {
        #region Protected Methods

        /// <summary>
        /// Добавляет форматированную строку с описанием вторичных процессов этапа в <paramref name="builder"/>.
        /// </summary>
        /// <param name="builder">Конструктор строки.</param>
        /// <param name="context"><inheritdoc cref="IStageTypeFormatterContext" path="/summary"/></param>
        /// <param name="settingsRows">Параметры этапа.</param>
        /// <param name="isClient">Признак формирования строки на клиенте.</param>
        protected static void AppendSecondaryProcessesNames(
            StringBuilder builder,
            IStageTypeFormatterContext context,
            IList? settingsRows,
            bool isClient)
        {
            if (!(settingsRows?.Count > 0))
            {
                return;
            }

            var stageRowID = context.StageRow.RowID;

            foreach (var settingsRow in settingsRows.Cast<IDictionary<string, object?>>())
            {
                if ((CardRowState) settingsRow.TryGet<int>(CardRow.SystemStateKey) == CardRowState.Deleted
                    || settingsRow.TryGet<Guid?>(KrConstants.StageRowIDReferenceToOwner) != stageRowID)
                {
                    continue;
                }

                AppendString(
                    builder,
                    settingsRow.TryGet<string>(KrConstants.KrForkSecondaryProcessesSettingsVirtual.SecondaryProcessName),
                    null,
                    true,
                    limit: isClient ? DefaultSettingMax : -1);
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krProcess/formatters/forkStageTypeFormatterBase.ts
import { Guid } from '@tessa/core';
import {
  CardRowState,
  IKrStageTypeFormatterContext,
  KrStageTypeFormatter,
  KrStageTypeFormattingHelper
} from '@tessa/platform';

export class ForkStageTypeFormatterBase extends KrStageTypeFormatter {
  //#region constants and static fields

  protected static readonly _secondaryProcessName = 'SecondaryProcessName';
  protected static readonly _krForkSecondaryProcessesSettingsVirtualSynthetic =
    KrStageTypeFormattingHelper.formatSectionName('KrForkSecondaryProcessesSettingsVirtual');
  protected static readonly _stageRowIDReferenceToOwner = 'StageRowID';

  //#endregion

  //#region protected methods

  protected appendSecondaryProcessesNames(
    builder: { text: string },
    context: IKrStageTypeFormatterContext
  ): void {
    const settingsRows = context.card.sections
      .tryGet(ForkStageTypeFormatterBase._krForkSecondaryProcessesSettingsVirtualSynthetic)
      ?.tryGetRows();

    if (settingsRows && settingsRows.length > 0) {
      const stageRowId = context.stageRow.rowId;

      for (const settingsRow of settingsRows) {
        const t = settingsRow.tryGetString(ForkStageTypeFormatterBase._stageRowIDReferenceToOwner);
        if (settingsRow.state === CardRowState.Deleted || !Guid.equals(t, stageRowId)) {
          continue;
        }

        this.appendString(
          builder,
          settingsRow.tryGetString(ForkStageTypeFormatterBase._secondaryProcessName),
          '',
          true,
          false,
          KrStageTypeFormatter.defaultSettingMax
        );
      }
    }
  }

  //#endregion
}
